src/Select/Select.tsx

```diff
@@ -1,78 +1,98 @@
-import React, { useState, cloneElement, ReactElement, useEffect, Ref } from 'react'
+import React, { useState, useEffect } from 'react'
 import styled from 'styled-components'
 
 import { BlockProps } from '../Block'
 import { Dropdown, DropdownPlacement } from '../Dropdown'
-import { InputProps } from '../Input/Input'
 import { SelectInput } from '../SelectInput'
 
-import { SelectListProps, SelectListData, SelectList } from '../SelectList'
+import { SelectListData, SelectList } from '../SelectList'
 
-interface InjectValueAndHandler {
-  onClick: () => void
-  name?: string
-  value?: string
-  placeholder?: string
-  ref?: React.Ref<HTMLInputElement>
-}
+const OptionsWrapper = styled.div<{ fullWidth: boolean }>`
+  width: ${({ fullWidth }) => fullWidth ? '100%' : 'auto'}
+`
 
-interface ChildProps extends Pick<SelectListProps, 'value' | 'data'> {
-  close: () => void
+const defaultInputValueAccessor = (item: any): string => {
+  if (!item && item !== 0) {
+    return ''
+  }
+  if (item.value) {
+    return defaultInputValueAccessor(item.value)
+  }
+  if (item.toString) {
+    return item.toString()
+  }
+  throw Error('Unexpected value. Custom `inputValueAccessor` should be defined')
 }
 
-export interface SelectProps extends Pick<SelectListProps, 'value' | 'data'> {
+interface InputComponentProps {
+  value?: string
   name?: string
   placeholder?: string
-  triggerComponent?: ReactElement<any>
-  onOutsideClick?: (isOpen: boolean, close: Function) => void
-  useTriggerComponentWidth?: boolean
-  onChange?: (value: any) => void
-  test?: string
-  children?: (props: ChildProps) => ReactElement
-}
-
-export const injectComponent = (component: ReactElement<any>, props: InjectValueAndHandler) => {
-  return cloneElement(component, {
-    ...props,
-    ...component.props
-  })
+  disabled?: boolean
+  onClick: () => void
+  onChange: () => void
 }
 
-const getTriggerComponent = (props: InjectValueAndHandler, ref: Ref<HTMLInputElement>, triggerComponent?: ReactElement<any>) => {
-  if (triggerComponent) return injectComponent(triggerComponent, { ...props, ref })
-  // Does not need onChange because SelectInput only display data
-  return <SelectInput {...props} ref={ref} />
+export interface SelectProps<V> {
+  value: V
+  inputValueAccessor?: (value: V) => string
+  InputComponent?: React.ComponentType<InputComponentProps>
+  data?: V extends SelectListData ? V[] : never
+  onChange?: V extends SelectListData ? ((item: V) => void) : never
+  name?: string
+  placeholder?: string
+  disabled?: boolean,
+  onOutsideClick?: (isOpen: boolean, close: () => void) => void
+  useInputComponentWidth?: boolean
+  children?: React.ReactElement
+  open?: boolean | null
+  onToggle?: (open: boolean) => void
 }
 
-const OptionsWrapper = styled.div<{ fullWidth: boolean }>`
-  width: ${({ fullWidth }) => fullWidth ? '100%' : 'auto'}
-`
-
-export const Select = ({
+export function Select <V = string> ({
```

Comment:
Here I would add a type parameter for the shape of the data. 

Right now, the type of `value` would still have to extend `SelectListData`, otherwise `data` would get typed as `never`. 

Given the `SelectProps` interface

```typescript
interface SelectProps<V> {
    value: V
    data?: V extends SelectListData ? V[] : never
}
````
If I give a string as a value it would reduce to this.

```typescript
interface SelectProps {
    value: string
    data?: string extends SelectListData ? string[] : never
}
````

which will reduce to this:

```typescript
interface SelectProps {
    value: string
    data?: never
}
````

If you have a separate `D` type argument then you can just say:

```typescript
interface SelectProps<V, D> {
    value: V
    inputValueAcessor: (value: V) => string
    data?: D
    onChange: (item: D) => void
}

```

Reply:
I also had something like that in my mind, but realized that second parameter is redundant. Anyways, if you provide `data`, the `value` should have the same shape as `data` has.
So the dependency exactly opposite, `data` depends on `value`. That's why there is conditional typing. If value, is just string (which is default), user should not provide any data, because it just will break the default implementation.

Reply:
In that case you would do it like this you mean:

```jsx
<Select>
  <Option/>
  <Option/>
  <Option/>
<Select/>
```

Reply:
yep, exactly. That's just extra guard, which says "don't give me any data, if you are using string as value"